Expand CI coverage and move actions into script

krcroft opened the pull request:
This PR makes a couple changes to the CI workflow:

- Adds more compiler coverage:
    - gcc to  MacOS (see note below)
    - 32 and 64bit gcc and clang to Windows
- With more builds, this separates them into per-OS workflow YAMLs (laying the foundation for more build environments: BSD? DOS? ... )
- Moves all functional commands from GitHub-syntax-YAML into scripts, which (besides eliminating repeated code), now serve a dual-purpose of being runnable outside of GitHub.
    - One script takes care of listing dependent packages for the given runtime environment
    - Another script takes care of configuring and building
    - These scripts can be leveraged by a nightly build & asset generator (TBD)
  
_Note_: adding GCC to MacOS is now "correct" from a build perspective, however to keep this PR focussed on the CI workflow I have not included the coreMIDI / AppleBlocks code-fixes here (so for now, the gcc macOS builds will fail; we will merge the coreMIDI / AppleBlocks later depending on how upstream wants to handle it).

dreamer commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,137 @@
+        run: C:\tools\msys64\usr\bin\bash -lc "cd d:/a/dosbox-staging/dosbox-staging; ./scripts/build.sh --bit-depth 32 --fdo --bin-path /mingw32/bin"
+
+      - name: Summarize warnings
+        run: C:\tools\msys64\usr\bin\python3 ./scripts/count-warnings.py build.log
```
There is no need to specify python3 inside msys2 in here - python3 included with the VM will work just as well: `run: python3 scripts\count-warnings.py build.log`

krcroft replied:
Just straight 'python3' didn't work with powershell. 
Adding `shell: bash` was no help either:
```
Run python3 ./scripts/count-warnings.py build.log
  python3 ./scripts/count-warnings.py build.log
  shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}
  env:
    CHERE_INVOKING: yes
d:\a\_temp\798b2c42-25f6-4255-b7a8-a55bb9e180f0.sh: line 1: python3: command not found
```
Falling back to the full path for now.

dreamer commented on scripts/list-build-dependencies.sh:
```diff
@@ -0,0 +1,298 @@
+#!/bin/bash
```
I imagined this file would be much smaller! :)

I suspect a file, which would look sort-of like this:

    #!/bin/bash
    
    case "$1" in
    	macos)
    		echo "coreutils autogen autoconf automake pkg-config libpng sdl sdl_net opusfile speexdsp" ;;
    	ubuntu)
    		echo "package1 package2 (...)" ;;
    	ubuntu-32)
    		echo "packageX packageY (...)" ;;
    	fedora)
    		echo "packageN packageM (...)" ;;
    	msys2-64)
    		echo "(...)" ;;
    	msys2-32)
    		echo "(...)" ;;
    	vcpkg)
    		echo "(...)" ;;
    esac


could probably fulfil all our requirements here? (of course this is not exact code, just an idea) I think we could drop compilers from being included with the build dependencies (those are usually not debatable for packaging in distros anyway, and could be easily hardcoded in the .yml files).

krcroft replied:
Yes! I think we can rip out all functions and boil it down to something elegant like that, and still handle stuff like --bit-depth 32 vs. 64 with only a couple if/elses. 
For now though, because it works and has been test maybe we could do this as an "version 2" optimization, and open an issue against it after merging?

dreamer replied:
OK :) I just checked and there are no shellcheck warnings in this file any more for this review.

dreamer commented on scripts/build.md:
```diff
@@ -0,0 +1,115 @@
+1. (optional) If you're interested in a specific branch:
+   * enter the repo directory: `cd dosbox-staging`
+   * list the available branches: `git branch --all`
+   * switch to a specific branch: `git checkout NAME_OF_BRANCH`
```
Overall this file is great, thanks for documenting this! The only comment from my side is about this section - I don't think it's necessary. We're not in a business of writing Git tutorials here an and inexperienced user might decide that `git checkout remotes/origin/real/branch-name` is a good idea ;)

dreamer commented on scripts/build.sh:
```diff
@@ -0,0 +1,554 @@
+#!/bin/bash
```
I still need to go read through this file more thoroughly (overall seems ok, but I think there are 2-3 places where script could be a bit simplified), but for a start: ShellCheck found several issues in this file, so I suggest installing it locally and fixing them (most issues are trivial).

I need to add shellcheck for bash scripts and pylint for python scripts in the future...

krcroft commented:
Regarding the optimization files - yes, I also despise binary files bloating repositories. Snipped.
If we can host them as an asset here, then the build script could wget/curl them.

I envision them rarely changing - I played through about 12 DOS games (which must be done in a single instance of DOSBox), and tried to hit all of the major functional paths (core types, video types, scalars, and audio card types) which generated over 100MB of kernel profile data, which gets baked down to those two portable ~300KB feedback files for gcc and clang.  

ideally I'd like to write a little "how to" or helper-script, and open it up to keen users on Vogons help to generate a more extensive profile. The workflow would be: run this command or script, start DOSBox and play games hitting these X, Y, Z functional paths, stop DOSBox, and run another command/script.

dreamer commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,142 @@
+        run:  C:\tools\msys64\usr\bin\bash -lc "./scripts/log-env.sh"
+
+      - name: Install dependencies
+        run:  C:\tools\msys64\usr\bin\bash -lc "./scripts/list-build-dependencies.sh --bit-depth 32 | xargs pacman -S --noconfirm"
```
Oh, I like new version of this line much more now :)

dreamer commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,142 @@
+        run:  C:\tools\msys64\usr\bin\bash -lc "./scripts/build.sh --compiler clang --bin-path /mingw64/bin"
+
+      - name: Summarize warnings
+        run:  C:\tools\msys64\usr\bin\python3 ./scripts/count-warnings.py build.log
```
I still think we don't need to route this step via python3 inside msys, but it's not a blocker.

krcroft replied:
I don't know why python3 isn't showing up in the PATH; I didn't look - maybe your original build.sh augmented PATH in the YAML (which I don't have)?

dreamer replied:
Ah, my bad, I looked at results of wrong job in my test commit 65be9fc4da348a021fedb

Following line will work:

      - name: Summarize warnings
        run: .\scripts\count-warnings.py build.log

I think at some point in recent past the default shell on Windows CI machines changed (at least I remember reading in the documentation it was `cmd` and now it says `powershell` is the default).

krcroft replied:
Will give it a whirl!  

Great to see powershell (and Windows in general) being able to launch scripts directly now (does it actually parse the first #! line or does it look at the .py extension?).

krcroft replied:
Confirmed working!
All builds are passing now; although details in the macOS clang build appears it didn't make it to the count-warnings step - so I'm running it again.

dreamer commented:
> Regarding the optimization files - yes, I also despise binary files bloating repositories. Snipped.
> If we can host them as an asset here, then the build script could wget/curl them.
> 
> I envision them rarely changing - I played through about 12 DOS games (which must be done in a single instance of DOSBox), and tried to hit all of the major functional paths (core types, video types, scalars, and audio card types) which generated over 100MB of kernel profile data, which gets baked down to those two portable ~300KB feedback files for gcc and clang.
> 
> ideally I'd like to write a little "how to" or helper-script, and open it up to keen users on Vogons help to generate a more extensive profile. The workflow would be: run this command or script, start DOSBox and play games hitting these X, Y, Z functional paths, stop DOSBox, and run another command/script.

Documentation and a script would certainly be helpful here - I was thinking about maybe setting separate repo, that would use git-lfs to store some demos / shareware versions of DOS games and could be used to automate regression testing - generating profile data out of that (or storing it in that dedicated repo) would be ok, I think.

krcroft commented:
Thinking about the profile data; it really doesn't make sense for it to live in any SCM; similar to release binaries..
GitLab support hosting release binaries - maybe that's an option? (even though it would be not ideal mixing and matching repo servers); although if you think it's reasonable to put them along side the demo games, that sounds workable too.

dreamer commented:
> Thinking about the profile data; it really doesn't make sense for it to live in any SCM; similar to release binaries..
> GitLab support hosting release binaries - maybe that's an option? (even though it would be not ideal mixing and matching repo servers); although if you think it's reasonable to put them along side the demo games, that sounds workable too.

Yes, GitLab is so much better than GitHub when it comes to hosting build results/artifacts/website for a project. I use it extensively for hosting packages built for https://gitlab.com/luxtorpeda/luxtorpeda project. Such a shame it does not handle cross-platform CI in free tier :(

For the record, I've set-up automatic mirroring of content of dosbox-staging for myself at https://gitlab.com/dreamer-tan/dosbox-staging (just in case) - we could easily add `.gitlab-ci.yml` file to the repo to trigger jobs at GitLab CI, but it's really nice to have all CI in one place.

krcroft commented on .github/workflows/macos.yml:
```diff
@@ -0,0 +1,72 @@
+      - name: Build
+        run: ./scripts/build.sh --compiler clang --lto --fdo
+      - name: "Summarize warnings"
+        run: python3 ./scripts/count-warnings.py build.log
```
Thanks; I caught the last one in macos.yml; pushing and testing now.

krcroft replied:
/Users/runner/runners/2.159.2/work/_temp/07a15170-f7bc-4b43-8040-eab14682dd52.sh: ./scripts/count-warnings.py: /usr/bin/python3: bad interpreter: No such file or directory

Apple.. causing unnecessary head-aches :face_with_head_bandage: 
Will revert.

dreamer replied:
Yeah, it's surprising… but that explains why I still see users on StackOverflow recommending hashbang `#!/usr/bin/env python3`. (Such hashbang caused issues to Linux distributions during python2 -> python3 transition and I personally saw it being a cause of user-bugs that couldn't be detected during testing by software developers - but it might have valid usecases in some contexts).

dreamer commented:
macOS: `./scripts/count-warnings.py: /usr/bin/python3: bad interpreter: No such file or directory` - I did not expect that :/ Scratch my request about about this build step on macOS then.

krcroft commented:
Excellent! Will drop in the exact warning and bug counts from this last CI run, and then will squash it down to one commit.

krcroft commented:
Thanks again for the guidance @dreamer 
 - Bug count shrunk to 99
 - Warning count  shrunk to 420
 - Rebased against master
 - Squashed to one commit titled and documented per first comment here

dreamer commented:
Merged, thanks! :)
